src/EditorFeatures/Core/Implementation/Classification/SyntacticClassificationTaggerProvider.TagComputer.cs

```diff
@@ -176,422 +227,317 @@ public void DisconnectFromWorkspace()
+                        }
+                    }
                 }
-            }
 
-            private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs args)
-            {
-                // We're getting an event for a workspace we already disconnected from
-                if (args.NewSolution.Workspace != _workspace)
+                void AddLexicalClassifications(SnapshotSpan span)
                 {
-                    // we are async so we are getting events from previous workspace we were associated with
-                    // just ignore them
-                    return;
+                    this.AssertIsForeground();
+
+                    classificationService.AddLexicalClassifications(
+                        span.Snapshot.AsText(), span.Span.ToTextSpan(), classifiedSpans, CancellationToken.None);
                 }
 
-                switch (args.Kind)
+                void AddSyntacticClassificationsForDocument(SnapshotSpan span, Document document, SyntaxNode? root)
                 {
-                    case WorkspaceChangeKind.ProjectChanged:
+                    this.AssertIsForeground();
+
+                    if (!_lastLineCache.TryUseCache(span, out var tempList))
+                    {
+                        tempList = ClassificationUtilities.GetOrCreateClassifiedSpanList();
+
+                        if (root != null)
                         {
-                            var documentId = _workspace.GetDocumentIdInCurrentContext(_subjectBuffer.AsTextContainer());
-                            if (documentId == null || documentId.ProjectId != args.ProjectId)
-                            {
-                                break;
-                            }
-
-                            var oldProject = args.OldSolution.GetProject(args.ProjectId);
-                            var newProject = args.NewSolution.GetProject(args.ProjectId);
-
-                            // make sure in case of parse config change, we re-colorize whole document. not just edited section.
-                            var configChanged = !object.Equals(oldProject.ParseOptions, newProject.ParseOptions);
-                            EnqueueProcessSnapshot(newProject.GetDocument(documentId));
-                            break;
+                            classificationService.AddSyntacticClassifications(document.Project.Solution.Workspace, root, span.Span.ToTextSpan(), tempList, cancellationToken);
                         }
-
-                    case WorkspaceChangeKind.DocumentChanged:
+                        else
                         {
-                            ProcessIfThisDocument(args.NewSolution, args.DocumentId);
-                            break;
+                            classificationService.AddSyntacticClassificationsAsync(document, span.Span.ToTextSpan(), tempList, cancellationToken).Wait(cancellationToken);
                         }
```

Comment:
new change. if we do have access to syntax, we can make these calls with no async and no blocking